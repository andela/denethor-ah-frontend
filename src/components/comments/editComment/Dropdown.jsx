import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class Dropdown extends Component {
  componentDidMount () {
    document.addEventListener('click', this.handleToggle);
  }

  componentWillUnmount () {
    document.removeEventListener('click', this.handleToggle);
  }

  handleToggle = () => {
    this.props.toggleModal();
  };

  render() {
    const { isAuthor, toggleMode, showHistory, hasHistory } = this.props;

    return (
      <div className='comment__edit-comment__drop-down' ref={node => this.modal = node}>
        <ul>
          {isAuthor && <li><button onClick={() => { toggleMode('edit') }}>Edit</button></li>}
          {
            hasHistory
              && (
                <li>
                  <button onClick={() => showHistory(true)}>
                    View History
                  </button>
                </li>
              )
          }
        </ul>
      </div>
    );
  }
}

Dropdown.propTypes = {
  isAuthor: PropTypes.bool,
  toggleModal: PropTypes.func,
  toggleMode: PropTypes.func,
  showHistory: PropTypes.func,
  hasHistory: PropTypes.object
}

export default Dropdown;
