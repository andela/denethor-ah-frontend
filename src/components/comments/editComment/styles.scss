.comment__edit-comment {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-right: 0.5rem;
  > p {
    opacity: 0.5;
  }
  svg {
    color: lighten(#07c, 20%);
    cursor: pointer;
  }
  .comment__edit-comment__drop-down {
    background: #f4f4f4;
    box-shadow: 1px 1px 2.5px #ddd, -1px -1px 2.5px #ddd;
    border-radius: 0.25rem;
    top: 1.5rem;
    right: -25%;
    position: absolute;
    width: 8rem;
    &:after {
      content: "";
      width: 0px;
      height: 0px;
      position: absolute;
      border-left: 10px solid transparent;
      border-right: 10px solid transparent;
      border-top: 10px solid transparent;
      border-bottom: 10px solid #f4f4f4;
      right: 25%;
      top: -18px;
    }
    ul {
      list-style-type: none;
      padding: 0;
      li {
        padding: 0 0.5rem;
        &:last-child {
          border-top: 1px solid #E0DFDF
        }
        &:first-child {
          border-top: none;
        }
        button {
          background: transparent;
          border: none;
          display: block;
          width: 100%;
          font: inherit;
        }
      }
    }
  }
}
