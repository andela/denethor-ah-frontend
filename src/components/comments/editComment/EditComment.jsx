import React, { Component } from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEllipsisH } from '@fortawesome/free-solid-svg-icons';
import Dropdown from './Dropdown';
import CommentEntries from '../CommentEntries';
import './styles.scss';
 
const customStyles = {
  content : {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)'
  }
};

export class EditComment extends Component {
  state = {
    showModal: false,
    showHistory: false
  }

  toggleModal = () => {
    this.setState(({ showModal }) => ({ showModal: !showModal }))
  }

  showHistory = (showHistory) => {
    this.setState({ showHistory });
  }

  render() {
    const { showModal, showHistory } = this.state;
    const { toggleMode, id, comment: { commentHistories = [], user: { id: commenterId } = {} } } = this.props;
    const isAuthor = id === commenterId;

    return (
      <div className='comment__edit-comment'>
        <Modal
          isOpen={showHistory}
          onRequestClose={() => this.showHistory(false)}
          style={customStyles}
          contentLabel="Example Modal"
          ariaHideApp={false}
        >
          <CommentEntries comments={commentHistories} />
        </Modal>

        <p>{commentHistories[0] && 'edited'}</p>
        {
          (isAuthor || commentHistories[0])
            && <FontAwesomeIcon icon={faEllipsisH} onClick={this.toggleModal} />
        }
        {showModal && <Dropdown
          toggleModal={this.toggleModal}
          isAuthor={isAuthor}
          toggleMode={toggleMode}
          showHistory={this.showHistory}
          hasHistory={commentHistories[0]}
          />}
      </div>
    );
  }
}

EditComment.propTypes = {
  comment: PropTypes.object,
  id: PropTypes.string,
  toggleMode: PropTypes.func,
  showHistory: PropTypes.func
};

const mapStateToProps = ({ profile: { id } }) => ({ id });

export default connect(mapStateToProps)(EditComment);
