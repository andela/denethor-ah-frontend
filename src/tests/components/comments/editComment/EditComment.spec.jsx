import React from 'react';
import { shallow } from 'enzyme';
import { EditComment } from '../../../../components/comments/editComment/EditComment';

let wrapper;

it('Should render static component', () => {
  const toggleMode = jest.fn();
  const comment = {
    commentHistories: []
  };
  const props = {
    toggleMode, comment, isAuthor: true, hasHistory: true
  }
  wrapper = shallow(<EditComment {...props} />);
  expect(wrapper).toMatchSnapshot();
});

it('Should call handle functions onclick', () => {
  
  wrapper.find('FontAwesomeIcon').simulate('click');
  expect(wrapper.state('showModal')).toBe(true);
});
