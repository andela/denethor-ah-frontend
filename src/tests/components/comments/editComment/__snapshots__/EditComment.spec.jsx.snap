// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<div
  className="comment__edit-comment"
>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Example Modal"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
      }
    }
  >
    <CommentEntries
      comments={Array []}
    />
  </Modal>
  <p />
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f141",
          "M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z",
        ],
        "iconName": "ellipsis-h",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    onClick={[Function]}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    symbol={false}
    title=""
    transform={null}
  />
</div>
`;
